### PR TITLE
iDeep: F.batch_normalization

### PR DESCRIPTION
**Pull request by niboshi**
iDeep support in F.batch_normalization (Separated from #3865. Based on #4276.)

### CODE REVIEW

**Comment by niboshi** [comment]
Rebased.

File: chainer/functions/normalization/batch_normalization.py [hunk 1]
```diff
@@ -55,15 +56,58 @@ def forward(self, inputs):
...
+                    self.decay, (1 - self.decay), self.mean)
+                self.running_var.inplace_axpby(
+                    self.decay, (1 - self.decay), self.var * adjust)
+            else:
+                self.running_mean *= self.decay
+                self.running_mean += self.mean * (1 - self.decay)
+                self.running_var *= self.decay
+                self.running_var += self.var * adjust * (1 - self.decay)
+
+            # ndarray ?
```
Comment (beam2d) [inline chainer/functions/normalization/batch_normalization.py]: Maybe you can remove this comment? (I'm not sure what it meant)

File: chainer/functions/normalization/batch_normalization.py [hunk 2]
```diff
@@ -300,7 +383,35 @@ def forward(self, inputs):
...
+
+            y, = intel64.ideep.batchNormalization.Forward(
+                intel64.ideep.array(x),
+                intel64.ideep.array(W),
+                intel64.ideep.array(mean),
+                intel64.ideep.array(var),
+                self.eps
+            )
+
+            # ndarray ?
```
Comment (beam2d) [inline chainer/functions/normalization/batch_normalization.py]: ditto

File: chainer/functions/normalization/batch_normalization.py [hunk 3]
```diff
@@ -55,15 +56,58 @@ def forward(self, inputs):
...
+                    intel64.ideep.array(W),
+                    None,
+                    None,
+                    self.eps
+                ))
 
-        if self.use_cudnn:
+            m = x.size // gamma.size
+            adjust = m / max(m - 1., 1.)
+            if (isinstance(self.running_mean, intel64.ideep.mdarray)
```
Comment (beam2d) [inline chainer/functions/normalization/batch_normalization.py]: Can we separate this if branch into one for `running_mean` and the other for `running_var`?

Comment (niboshi) [reply chainer/functions/normalization/batch_normalization.py]: Fixed.

File: chainer/functions/normalization/batch_normalization.py (hunk 2 above)
Comment (niboshi) [reply chainer/functions/normalization/batch_normalization.py]: Fixed.

File: chainer/functions/normalization/batch_normalization.py (hunk 1 above)
Comment (niboshi) [reply chainer/functions/normalization/batch_normalization.py]: Fixed.

**Comment by beam2d** [comment]
Jenkins, test this please.

**Pull request merged**